File: dev-server.py
#!/usr/bin/env python3
# dev-server.py  （零依赖，仅用 Python 标准库）
# 功能同 dev-server.mjs：静态托管 + 可选 /proxy/<vendor>/<path> 反向代理。
# 说明：Python 版代理为“整块缓冲”转发（非增量流式），预览足够；如需打字机流式效果请用 Node 版。
#
# 安全约定（与 dev-server.mjs 一致）：
# - 只绑定 127.0.0.1，密钥文件/.git/node_modules 不经静态服务暴露。
# - 代理只接受本机来源，拒绝跨站页面 CSRF 驱动。

import http.server
import socketserver
import urllib.request
import urllib.error
import urllib.parse
import json
import os
import re

ROOT = os.path.dirname(os.path.abspath(__file__))
HOST = os.environ.get('HOST', '127.0.0.1')
PORT = int(os.environ.get('PORT', '5173'))

VENDOR_BASE = {
    'openai': 'https://api.openai.com/v1',
    'anthropic': 'https://api.anthropic.com/v1',
    'gemini': 'https://generativelanguage.googleapis.com/v1beta',
    'ollama': 'http://localhost:11434',
}

secrets = {}
sp = os.path.join(ROOT, 'preview', 'secrets.json')
if os.path.exists(sp):
    try:
        secrets = json.load(open(sp, encoding='utf-8'))
    except Exception:
        secrets = {}

MIME = {
    '.html': 'text/html; charset=utf-8',
    '.js': 'text/javascript; charset=utf-8',
    '.css': 'text/css; charset=utf-8',
    '.json': 'application/json; charset=utf-8',
}

# 静态服务永不下发的路径
STATIC_DENYLIST = ['secrets.json', '.git', 'node_modules']


def _denied(urlpath):
    p = urlpath.lower()
    return any(d in p for d in ('/' + x for x in STATIC_DENYLIST))


def _local_origin(headers):
    host = (headers.get('Host') or '').split(':')[0]
    if host not in ('localhost', '127.0.0.1', '[::1]', '::1'):
        return False
    origin = headers.get('Origin')
    if not origin:
        return True
    try:
        o = urllib.parse.urlparse(origin)
        return o.hostname in ('localhost', '127.0.0.1', '::1')
    except Exception:
        return False


class Handler(http.server.BaseHTTPRequestHandler):
    def _serve_static(self):
        urlpath = urllib.parse.unquote(self.path.split('?')[0], errors='replace')
        if urlpath == '/':
            urlpath = '/preview/index.html'
        if _denied(urlpath):
            self.send_error(403); return
        fp = os.path.normpath(os.path.join(ROOT, urlpath.lstrip('/')))
        # normpath 后必须是 ROOT 或其子路径；startsWith 裸比较会被同前缀兄弟目录绕过
        if fp != ROOT and not fp.startswith(ROOT + os.sep):
            self.send_error(403); return
        if not os.path.isfile(fp):
            self.send_error(404); return
        with open(fp, 'rb') as f:
            data = f.read()
        self.send_response(200)
        self.send_header('Content-Type', MIME.get(os.path.splitext(fp)[1], 'application/octet-stream'))
        self.send_header('Content-Length', str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def _proxy(self, vendor, rest):
        base = VENDOR_BASE.get(vendor)
        if not base:
            self.send_error(400, 'unknown vendor'); return
        if not _local_origin(self.headers):
            self.send_error(403, 'proxy only serves local origins'); return
        target = base + '/' + rest
        if vendor == 'gemini' and secrets.get('gemini'):
            target = re.sub(r'([?&]key=)[^&]*', r'\1' + secrets['gemini'], target)
            if 'key=' not in target:
                target += ('&' if '?' in target else '?') + 'key=' + secrets['gemini']
        length = int(self.headers.get('Content-Length', 0) or 0)
        body = self.rfile.read(length) if length else None
        headers = {}
        if self.headers.get('Content-Type'):
            headers['Content-Type'] = self.headers['Content-Type']
        if vendor == 'openai' and secrets.get('openai'):
            headers['Authorization'] = 'Bearer ' + secrets['openai']
        if vendor == 'anthropic' and secrets.get('anthropic'):
            headers['x-api-key'] = secrets['anthropic']
            headers['anthropic-version'] = '2023-06-01'
        if vendor == 'gemini':
            headers.pop('Authorization', None)
        req = urllib.request.Request(target, data=body, headers=headers, method=self.command)
        try:
            with urllib.request.urlopen(req, timeout=120) as resp:
                data = resp.read()
                self.send_response(resp.status)
                self.send_header('Content-Type', resp.headers.get('Content-Type', 'application/json'))
                self.send_header('Content-Length', str(len(data)))
                self.end_headers()
                self.wfile.write(data)
        except urllib.error.HTTPError as e:
            err = e.read()
            self.send_response(e.code)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Content-Length', str(len(err)))
            self.end_headers()
            self.wfile.write(err)
        except Exception as e:
            self.send_error(502, str(e))

    def do_GET(self):
        m = re.match(r'^/proxy/(\w+)/(.*)$', self.path)
        if m:
            self._proxy(m.group(1), m.group(2))
        else:
            self._serve_static()

    def do_POST(self):
        m = re.match(r'^/proxy/(\w+)/(.*)$', self.path)
        if m:
            self._proxy(m.group(1), m.group(2))
        else:
            self._serve_static()

    def log_message(self, *a):
        pass


if __name__ == '__main__':
    # 用多线程服务器：流式代理请求较慢（最长 120s）时，仍可同时响应页面静态 JS/CSS，
    # 避免对话期间预览页“卡住加载”。（与 .mjs 版的异步并发行为对齐）
    with socketserver.ThreadingTCPServer((HOST, PORT), Handler) as s:
        print(f'\n  本地预览已启动:  http://localhost:{PORT}（仅监听 {HOST}）')
        print(f'  代理模式:       {"已配置密钥" if secrets else "未配置 preview/secrets.json（仅直连模式）"}\n')
        s.serve_forever()

File: dev-server.mjs
// dev-server.mjs  （零依赖，仅用 Node 内置模块）
// 1) 静态托管整个项目（含 preview/ 与核心模块）
// 2) 可选反向代理：/proxy/<vendor>/<path> -> 真实厂商接口，密钥由服务端注入
//    用于规避官方 API 的浏览器 CORS，且 API Key 不进入前端。
//
// 安全约定（本地开发服务也必须遵守）：
// - 只绑定 127.0.0.1：默认监听会绑到 0.0.0.0，同网段任何设备都能拿到静态文件
//   并通过 /proxy/* 白嫖本机配置的 API 配额。
// - 密钥文件（preview/secrets.json）、.git、node_modules 永不经静态服务暴露。
// - 代理只接受本机来源：浏览器发起的跨站请求（恶意网页 CSRF 驱动带密钥的代理）会被拒绝。

import http from 'node:http';
import https from 'node:https';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname);
const HOST = process.env.HOST || '127.0.0.1';
const PORT = Number(process.env.PORT) || 5173;

// 各厂商真实 Base（代理模式使用）
const VENDOR_BASE = {
  openai: 'https://api.openai.com/v1',
  anthropic: 'https://api.anthropic.com/v1',
  gemini: 'https://generativelanguage.googleapis.com/v1beta',
  ollama: 'http://localhost:11434',
  ima: 'https://ima.qq.com',
};

// 服务端密钥（绝不发送到前端）。从 preview/secrets.json 读取。
let secrets = {};
try {
  secrets = JSON.parse(fs.readFileSync(path.join(ROOT, 'preview', 'secrets.json'), 'utf8'));
} catch { /* 直连模式无需该文件 */ }

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.ico': 'image/x-icon',
};

// 静态服务永不下发的路径：密钥文件、版本库、依赖目录
const STATIC_DENYLIST = ['secrets.json', '.git', 'node_modules'];

function isDenied(urlPath) {
  const p = urlPath.toLowerCase();
  return STATIC_DENYLIST.some(d => p === '/' + d || p.startsWith('/' + d + '/') || p.includes('/' + d + '/'));
}

function serveStatic(req, res) {
  let urlPath;
  try {
    urlPath = decodeURIComponent(req.url.split('?')[0]);
  } catch {
    // 畸形编码（/%ZZ）会让 decodeURIComponent 抛 URIError；不捕获会打崩整个进程
    res.writeHead(400); res.end('bad request'); return;
  }
  if (urlPath === '/') urlPath = '/preview/index.html';
  if (isDenied(urlPath)) { res.writeHead(403); res.end('forbidden'); return; }
  const filePath = path.resolve(path.join(ROOT, urlPath));
  // path.resolve + 分隔符后缀比较：startsWith(ROOT) 会被同前缀兄弟目录
  // （如 Browser_AI_Extensions_backup）绕过
  if (filePath !== ROOT && !filePath.startsWith(ROOT + path.sep)) { res.writeHead(403); res.end('forbidden'); return; }
  fs.readFile(filePath, (err, data) => {
    if (err) { res.writeHead(404); res.end('not found'); return; }
    res.writeHead(200, { 'Content-Type': MIME[path.extname(filePath)] || 'application/octet-stream' });
    res.end(data);
  });
}

function filterHeaders(h) {
  const out = {};
  for (const [k, v] of Object.entries(h)) {
    const lk = k.toLowerCase();
    if (['content-length', 'connection', 'transfer-encoding'].includes(lk)) continue;
    out[k] = v;
  }
  return out;
}

/** 代理只服务本机来源：拒绝跨站页面驱动的请求（CSRF 驱动本机代理打厂商接口） */
function allowedOrigin(req) {
  const host = String(req.headers.host || '').split(':')[0];
  if (host !== 'localhost' && host !== '127.0.0.1' && host !== '[::1]') return false;
  const origin = req.headers.origin;
  if (!origin) return true; // curl / 同源 GET 等无 Origin 头
  try {
    const o = new URL(origin);
    return o.hostname === 'localhost' || o.hostname === '127.0.0.1' || o.hostname === '[::1]';
  } catch {
    return false;
  }
}

function proxy(req, res, vendor, restPath) {
  const base = VENDOR_BASE[vendor];
  if (!base) { res.writeHead(400); res.end('unknown vendor: ' + vendor); return; }
  if (!allowedOrigin(req)) {
    res.writeHead(403, { 'Access-Control-Allow-Origin': 'null' });
    res.end('forbidden: proxy 仅接受本机来源请求');
    return;
  }

  let target = base + '/' + restPath;
  // Gemini：把 key 参数替换为服务端密钥
  if (vendor === 'gemini' && secrets.gemini) {
    target = target.replace(/([?&]key=)[^&]*/, '$1' + secrets.gemini);
    if (!/[?&]key=/.test(target)) {
      target += (target.includes('?') ? '&' : '?') + 'key=' + secrets.gemini;
    }
  }

  const parsed = new URL(target);
  const chunks = [];
  req.on('data', (c) => chunks.push(c));
  req.on('end', () => {
    const body = Buffer.concat(chunks);
    const headers = { ...req.headers };
    delete headers.host; delete headers.connection; delete headers['content-length'];
    delete headers.origin; delete headers.referer;
    if (vendor === 'openai' && secrets.openai) headers.authorization = 'Bearer ' + secrets.openai;
    if (vendor === 'anthropic' && secrets.anthropic) {
      headers['x-api-key'] = secrets.anthropic;
      headers['anthropic-version'] = '2023-06-01';
    }
    if (vendor === 'gemini') delete headers.authorization;
    if (vendor === 'ima' && secrets.ima) {
      headers['ima-openapi-clientid'] = secrets.ima.clientId || '';
      headers['ima-openapi-apikey'] = secrets.ima.apiKey || '';
    }

    const lib = parsed.protocol === 'https:' ? https : http;
    const opt = {
      method: req.method,
      headers,
      hostname: parsed.hostname,
      port: parsed.port || (parsed.protocol === 'https:' ? 443 : 80),
      path: parsed.pathname + parsed.search,
    };
    const p = lib.request(opt, (up) => {
      res.writeHead(up.statusCode, filterHeaders(up.headers));
      up.pipe(res); // 流式透传（支持 SSE / 增量）
      up.on('error', () => { try { res.destroy(); } catch (_) {} });
    });
    p.on('error', (e) => {
      // 上游失败时可能响应头已发出（如中途断流）：再 writeHead 会抛 ERR_HTTP_HEADERS_SENT 打崩进程
      if (res.headersSent) { try { res.destroy(); } catch (_) {} return; }
      res.writeHead(502); res.end('proxy error: ' + e.message);
    });
    if (body.length) p.write(body);
    p.end();
  });
}

const server = http.createServer((req, res) => {
  try {
    const m = req.url.match(/^\/proxy\/(\w+)\/(.*)$/);
    if (m) { proxy(req, res, m[1], m[2]); return; }
    serveStatic(req, res);
  } catch (e) {
    // 兜底：任何请求处理异常都不允许带崩整个 dev server
    if (!res.headersSent) { try { res.writeHead(500); res.end('internal error'); } catch (_) {} }
    console.warn('[dev-server] request error:', e && e.message);
  }
});

server.listen(PORT, HOST, () => {
  console.log(`\n  本地预览已启动:  http://localhost:${PORT}（仅监听 ${HOST}）`);
  console.log(`  代理模式:       ${Object.keys(secrets).length ? '已配置密钥' : '未配置 preview/secrets.json（仅直连模式）'}\n`);
});
